### PR TITLE
Add documentation for types and serializers

### PR DESCRIPTION
**Pull request by elronbandel**
For: https://github.com/IBM/unitxt/issues/1283
Closes: https://github.com/IBM/unitxt/issues/1283

### CODE REVIEW

File: docs/docs/glossary.rst [hunk 1]
```diff
@@ -284,3 +284,12 @@ The verbalization involves verbalizing the task input fields for the input, and
...
    :width: 75%
    :align: center
+
+
+Serialization
+--------------
+
+Is the process of taking a specific field of a task a converting it into a string.
+This process is controlled by the :ref:`Serializer <serializers>`.
+Unlike the verbalization which is task specific, namely putting into words an input of a specific task, the serialization
```
Comment (yoavkatz) [inline docs/docs/glossary.rst:294]: I'm not sure this is distinction is right (verbalization is task specific vs serialization which I not).

**Verbalization** is the process of taking the task fields and converting them into their
textual representation, which is provided as input to the model.

Type serialization is part of the verbalization process and is often related to the template used.
The other type is the template formatting, which verbalizes how the text representation of the fields are arranged in the prompt.

File: docs/docs/types_and_serializers.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    dataset = load_dataset(
+        card=card,
+        template=InputOutputTemplate(
+            instruction="Summarize the following dialog.",
+            input_format="{dialog}",
+            output_format="{summary}",
+        ),
+        serializer=DialogSerializer(),
+    )
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:121]: I think we should show the expected output.  Also, we should show or mention how to set the serializer in the template.

File: docs/docs/types_and_serializers.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+Tags and Descriptions
+=====================================
+
+Unitxt employ different tools for serializing data into textual
+format. One of this mechanism is the Type-Serializers mechanism which assign serialization to specific types.
+For example consider the following ``typing types``:
+
+.. code-block:: python
+
+    Text = NewType("Text", str)
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:11]: What do I need to import for NewType?

Comment (elronbandel) [reply docs/docs/types_and_serializers.rst:11]: It's `typing` class. Will add its import.

File: docs/docs/types_and_serializers.rst (hunk 2 above)
Comment (elronbandel) [reply docs/docs/types_and_serializers.rst:121]: Will do.

File: docs/docs/glossary.rst [hunk 4]
```diff
@@ -284,3 +284,12 @@ The verbalization involves verbalizing the task input fields for the input, and
    :alt: The unitxt prompt structure
    :width: 75%
    :align: center
+
+
+Serialization
+--------------
+
+Is the process of taking a specific field of a task a converting it into a string.
```
Comment (yoavkatz) [inline docs/docs/glossary.rst:292]:
```suggestion
**Serialization** Is the process of taking a specific field of a task a converting it into a string.  
```

File: docs/docs/glossary.rst (hunk 1 above)
Comment (yoavkatz) [inline docs/docs/glossary.rst:294]:
```suggestion
The serialization of the task fields is part of the verbalization process performed by the  :ref:`Templates <template>`.
```

File: docs/docs/types_and_serializers.rst [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+Adding a Serializer to a Template
+------------------------------------
+
+Another option is to set a default serializer for a given template. When creating a template, we need to add all the serializers for all the types we want to support. For this purpose, we use a multi-type serializer that wraps all the serializers together.
+
+.. code-block:: python
+
+    from unitxt.serializers import (
+        MultiTypeSerializer, ImageSerializer, TableSerializer, DialogSerializer, ListSerializer,
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:140]:
```suggestion
        ListSerializer
```

File: docs/docs/types_and_serializers.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    )
+
+    serializer = MultiTypeSerializer(
+        serializers=[
+            ImageSerializer(),
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:145]:
```suggestion
       
```

File: docs/docs/types_and_serializers.rst [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+            TableSerializer(),
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:146]:
```suggestion
    
```

File: docs/docs/types_and_serializers.rst [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            DialogSerializer(),
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:147]:
```suggestion
            MyDialogSerializer(),
```

File: docs/docs/types_and_serializers.rst [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+.. code-block:: python
+
+    dataset = load_dataset(
+        card=card,
+        template=InputOutputTemplate(
+            instruction="Summarize the following dialog.",
+            input_format="{dialog}",
+            output_format="{summary}",
+        ),
+        serializer=DialogSerializer(),
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:119]:
```suggestion
        serializer=MyDialogSerializer(),
```

File: docs/docs/types_and_serializers.rst [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+Defining a Serializer for a Type
+---------------------------------
+
+Once the types are registered, we can define serializers for those types. For example, consider creating a serializer for the ``Dialog`` type:
+
+.. code-block:: python
+
+    from unitxt.serializers import SingleTypeSerializer
+
+    class DialogSerializer(SingleTypeSerializer):
```
Comment (yoavkatz) [inline docs/docs/types_and_serializers.rst:52]:
```suggestion
    class MyDialogSerializer(SingleTypeSerializer):
```

File: docs/docs/types_and_serializers.rst (hunk 9 above)
Comment (yoavkatz) [reply docs/docs/types_and_serializers.rst:119]: I use MyDialogSerializer() to highlight is a custom dialog.

File: docs/docs/examples.rst [hunk 11]
```diff
@@ -226,3 +226,16 @@ This approach can be adapted for various image-text to text tasks, such as image
...
 Related documentation: :ref:`Multi-Modality Guide <multi_modality>`.
+
+
+Types and Serializers
+----------------------------
+
+Custom Types and Serializers
++++++++++++++++++++++++++++++
+
+This example show how to define new data types as well as the way these data type should be handled when processed to text.
```
Comment (yoavkatz) [inline docs/docs/examples.rst:237]:
```suggestion
This example shows how to prompt a model when the task has fields of complex types such as dialogs, images, and tables.  It demonstrates how to define new data types and how these data types are converted to a textual representation in the prompt provided to the model.
```

File: docs/docs/glossary.rst (hunk 4 above)
Comment (welisheva22) [reply docs/docs/glossary.rst:292]: Thanks for asking for my help -- looking good, couple more comments
 - please change "Is" to "is" with a lowercase i
 - please change "a" to "and" here -- "...field of a task a converting it..." to "...field of a task and converting it..."

**Pull request merged**